Feature/fuzzy search logic

danielrosehill opened the pull request:
<!--
  This template provides some ideas of things to include in your PR description.
  To start, try providing a short summary of your changes in the Title above.
  If a section of the PR template does not apply to this PR, then delete that section.
 -->

## What type of PR is this?

_(REQUIRED)_

<!--
  Delete any of the following that do not apply:
 -->

- bug
- cleanup
- documentation
- feature

## What this PR does / why we need it:

_(REQUIRED)_

<!--
  What goal is this change working towards?
  Provide a bullet pointed summary of how each file was changed.
  Briefly explain any decisions you made with respect to the changes.
  Include anything here that you didn't include in *Release Notes*
  above, such as changes to CI or changes to internal methods.
-->

## Which issue(s) this PR fixes:

_(REQUIRED)_

<!--
If this PR fixes one of more issues, list them here.
One line each, like so:
Fixes #123
Fixes #39
-->

## Special notes for your reviewer:

_(fill-in or delete this section)_

<!--
   Is there any particular feedback you would / wouldn't like?
   Which parts of the code should reviewers focus on?
-->

## Testing

_(fill-in or delete this section)_

<!--
  Describe how you tested this change.
-->

thevortexcloud commented on frontend/pages/items.vue:
```diff
@@ -310,59 +322,43 @@
 
-    await search();
+    // Trigger a search with the reset values
+    search();
```
Why is this not awaited any more?

thevortexcloud reviewed (commented):
Not sure why you closed the last PR, but you did not answer my question. So I am going to put it here again:

Out of curiosity, what order are results returned in? To me it makes sense for exact matches to be returned before fuzzy matches. Fuzzy search that is too fuzzy is frustrating and I often find it useless. EG if I am searching for a USB A to A cable and it returns USB C results first then the search has failed as far as I am concerned.

I also see that you have added a toggle to turn the fuzzy search off in this PR, which does not really adddress the problem.

I have added a review comment. But none of this is up to me, I am not in charge around here.

danielrosehill commented:
> Not sure why you closed the last PR, but you did not answer my question. So I am going to put it here again:
> 
> Out of curiosity, what order are results returned in? To me it makes sense for exact matches to be returned before fuzzy matches. Fuzzy search that is too fuzzy is frustrating and I often find it useless. EG if I am searching for a USB A to A cable and it returns USB C results first then the search has failed as far as I am concerned.
> 
> I also see that you have added a toggle to turn the fuzzy search off in this PR, which does not really adddress the problem.
> 
> I have added a review comment. But none of this is up to me, I am not in charge around here.

Hey!

I'm a Github novice I'm afraid (well, not at Github, but at working on collaborative codebases here).

I closed the other PRs because I realised (sadly too late) that I was working off an outdated fork that was significantly behind the current project. Rather than get bogged down trying to fix individual conflicts, I thought it would be quicker to just close those out and open new ones.  I didn't see your comment - sorry (I was mostly looking for the test results and between all the bot comments I must have missed them).  Thank you for persevering and writing it again. 

## Implementation 

It's added under the options.

For example, keyword search for USB A with lowercase 
 
![image](https://github.com/user-attachments/assets/2ec89620-3e09-4779-b876-40b7eba5ce2b)

Toggling on the fuzzy search adds one result

![image](https://github.com/user-attachments/assets/d9c3360d-5d55-4eea-8f2e-63adae3b8940)

Generally I find that it adds just a couple more results based on minor discrepancies in spelling. 

In the previous PR, I added another search "enhancement" that widened the logic more significantly but also resulted in an unacceptable amount of false positives (sorry, I discovered this only in retrospect after trying it out for a few days, I guess another reason to close the PR!)

danielrosehill commented:
> Out of curiosity, what order are results returned in? To me it makes sense for exact matches to be returned before fuzzy matches. Fuzzy search that is too fuzzy is frustrating and I often find it useless. EG if I am searching for a USB A to A cable and it returns USB C results first then the search has failed as far as I am concerned.

Agree completely.

But I think that the new implementation is a good compromise:

By default, it's disabled; user can enable it when you know that the search is, for sure, missing something. 

Perhaps more ideally there would be a granularity setting but ... I think that would be overcomplicating things.

thevortexcloud commented:
> I'm a Github novice I'm afraid (well, not at Github, but at working on collaborative codebases here).

Fair enough. For future reference you could have technically done a force push to rewrite history (Just don't rewrite commits you did not make/have already been merged) and clean up the PR. As you have full control over your fork up until it gets merged. Although you would still have to merge all the missing commits into your fork somehow.

> Perhaps more ideally there would be a granularity setting but ... I think that would be overcomplicating things.

I personally think making it a toggle makes things more complicated, at least UX wise as there is already a lot of hidden options and adding another can be overwhelming for people (especially in your screenshot as there are now 7 options in that pop up, and presumably other people will add more later). This would probably fit better as a search operator. EG You could do `"USB" a` to exact search for `USB` and then fuzzy match on `a`. The default will be then to fuzzy match everything. Similar to how google search works. This would also build upon the existing search operators. You would however need to make some sort of parser for it and not break the existing operators (EG adding `#` currently does a search by asset id).


> The ordering logic isn't changed:

Excellent.
